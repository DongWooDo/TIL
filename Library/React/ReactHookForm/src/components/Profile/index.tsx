import { forwardRef } from 'react';

import {
  useForm,
  SubmitHandler,
  UseFormRegister, useFormContext, FormProvider,
} from 'react-hook-form';

import _ from 'lodash';

interface IFormValues {
  firstName: string;
  lastName: string;
  phoneNumber: string;
}

const Input = forwardRef<
  HTMLInputElement,
  {
    id: string;
    label: string;
    defaultValue: string;
    ariaInvalid: boolean;
    required: boolean;
  } & ReturnType<UseFormRegister<IFormValues>>
  >(({
    id, label, defaultValue, required, ariaInvalid,
    name, onChange, onBlur,
  }, ref) => (
    <div>
      <label htmlFor={id}>{label}</label>
      <input
        id={id}
        defaultValue={defaultValue}
        required={required}
        name={name}
        onChange={onChange}
        onBlur={onBlur}
        ref={ref}
        aria-invalid={ariaInvalid}
      />
    </div>
  ));

function DeepNest() {
  console.log('DeepNest rendering');
  const { register, formState: { errors } } = useFormContext();

  return (
    <div>
      <Input
        id="phone-number"
        label="휴대 전화"
        ariaInvalid={!!errors.phoneNumber}
        {...register('phoneNumber', { required: { value: true, message: '휴대 전화를 입력해주세요.' } })}
        required={false}
        defaultValue=""
      />
      <p role="alert">{errors?.phoneNumber?.message}</p>
    </div>
  );
}

export default function Profile() {
  console.log('Profile rendering');
  const methods = useForm<IFormValues>();

  // errors가 있을 경우 onSubmit 함수는 실행되지 않는다.
  const onSubmit: SubmitHandler<IFormValues> = (data) => {
    // { name: 'example', onBlur: fn, onChange: fn, ref: fn }
    // console.log(register('example'));

    console.log(data);
  };

  // { firstName: { message: '', ref: input, type: 'required'} }
  // console.log(errors);

  const { handleSubmit, formState: { errors }, register } = methods;

  return (
    <FormProvider {...methods}>
      <form onSubmit={handleSubmit(onSubmit)}>
        <Input
          id="first-name"
          label="First Name"
          defaultValue=""
          {...register('firstName', { required: { value: true, message: 'First Name is required' } })}
          ariaInvalid={!!errors.firstName}
          required
        />
        <p role="alert">{errors?.firstName?.message}</p>

        <Input
          id="last-name"
          label="Last Name"
          defaultValue=""
          {...register('lastName', { required: { value: true, message: 'Last Name is required' } })}
          ariaInvalid={!!errors.firstName}
          required={false}
        />
        <p role="alert">{_.get(errors, 'lastName.message')}</p>

        <DeepNest />

        <button type="submit">제출</button>
      </form>
    </FormProvider>
  );
}
