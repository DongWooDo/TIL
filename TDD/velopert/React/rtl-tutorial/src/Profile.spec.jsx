import { render } from "@testing-library/react";
import Profile from "./Profile";

describe("<Profile />", () => {
  it("matches snapshot", () => {
    const utils = render(
      <Profile username="vue만 하고 싶다" name="dongwoo"></Profile>
    );
    expect(utils.container).toMatchSnapshot();
  });

  it("shows the props correctly", () => {
    const utils = render(<Profile username="vue만 하고 싶다" name="dongwoo" />);
    utils.getByText("vue만 하고 싶다"); // vue만 하고 싶다 라는 텍스트를 가진 엘리먼트가 있는지 확인
    utils.getByText("(dongwoo)"); // (김민준) 이라는 텍스트를 가진 엘리먼트가 있는지 확인
    utils.getByText(/dong/); // 정규식 /김/ 을 통과하는 엘리먼트가 있는지 확인
  });
});
