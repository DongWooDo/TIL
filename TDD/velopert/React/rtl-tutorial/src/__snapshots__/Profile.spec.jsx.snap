// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Profile /> matches snapshot 1`] = `
<div>
  <div>
    <b>
      vue만 하고 싶다
    </b>
     
    <span>
      (
      dongwoo
      )
    </span>
  </div>
</div>
`;
